### ISSUE TITLE
Allow path-specific dependencies in [dev-dependencies]

### ISSUE THREAD

**Issue by lukaslueg** [rust-lang/cargo#3745]
Cargo currently refuses any `Cargo.toml` that includes a path-specific dependency. This is overly zealous, as crates.io could allow `dev-dependencies` to be path-specific: These never get touched while cargo builds something from crates.io.

A use case for a path-specific `dev-dependency` is a test-crate for integration tests, breaking a dependency cycle where the main crate dev-depends on a test-crate (as a path-specific `dev-dependency`) and the test-crate depends on the main-crate (as a normal `dependency`). Cargo actually does that right thing on `cargo test` and
1. compiles the main crate first (satisfying the test-crate's `dependency`)
2. the test-crate second (satisfying the `dev-dependency`) and
3. executing the tests

On a normal `cargo build`, the path-specific `dev-dependency` will be ignored.

CC @alexcrichton

**Comment by alexcrichton** [comment]
Yeah in general I think crates.io knows too much about dev-dependencies. I think we should be able to get by without verifying that dev-dependencies come from the same source or are present on crates.io itself.

**Comment by lukaslueg** [comment]
How to move forward here? Is it sufficient to just remove the locality check for `dev-dependency´?

**Comment by alexcrichton** [comment]
Unfortunately I have not thought through what this would entail. I don't know what it would take to solve it or whether a naive solution would break other code.

**Comment by ehuss** [comment]
I think this is now addressed by #7333.  If that is incorrect, feel free to reopen.
